modules/channel-smooch/src/backend/client.ts

```diff
@@ -0,0 +1,128 @@
+import * as sdk from 'botpress/sdk'
+import Smooch from 'smooch-core'
+
+import { Config } from '../config'
+
+import { Clients } from './typings'
+
+const name = 'smooch.sendMessage'
+
+export class SmoochClient {
+  private smooch: any
+  private path: string
+
+  constructor(
+    private bp: typeof sdk,
+    private botId: string,
+    private config: Config,
+    private router: any,
+    private route: string
+  ) {}
+
+  async initialize() {
+    this.smooch = new Smooch({
+      keyId: this.config.keyId,
+      secret: this.config.secret,
+      scope: 'app'
+    })
+
+    const url = (await this.router.getPublicPath()) + this.route
+    this.path = url.replace('BOT_ID', this.botId)
+
+    await this.smooch.webhooks.create({
+      target: this.path,
+      triggers: ['message:appUser']
+    })
+  }
+
+  async delete() {
+    const { webhooks } = await this.smooch.webhooks.list()
+    for (const hook of webhooks) {
+      if (hook.target == this.path) {
+        await this.smooch.webhooks.delete(hook._id)
+      }
+    }
+  }
+
+  async handleWebhookRequest(body: any) {
+    if (!body.messages) {
+      return
+    }
+
+    for (const message of body.messages) {
+      if (message.type !== 'text') {
+        continue
+      }
+
+      await this.bp.events.sendEvent(
+        this.bp.IO.Event({
+          botId: this.botId,
+          channel: 'smooch',
+          direction: 'incoming',
+          type: 'text',
+          payload: {
+            type: 'text',
+            markdown: false,
+            text: message.text
+          },
+          preview: message.text,
+          target: body.appUser._id
+        })
+      )
+    }
+  }
+
+  async handleOutgoingEvent(event: sdk.IO.Event, next: sdk.IO.MiddlewareNextCallback) {
+    if (event.type === 'typing') {
+      await this.smooch.appUsers.conversationActivity({
+        appId: this.smooch.keyId,
+        userId: event.target,
+        activityProps: {
+          role: 'appMaker',
+          type: 'typing:start'
+        }
+      })
+    } else if (event.type === 'text') {
```

Comment:
Only outgoing events supported are text & typing info. Should it be expanded to more complex types? Smooch connects to many different channels (including SMS) so support for these won't be consistent.

Reply:
It should at least support single choice & images, otherwise it's not very useful. Carousels are also commonly used. If possible, it should support all the builtin content types